### PR TITLE
Maximum size for a release

### PR DESCRIPTION
**Pull request by margaale**
#### Database Migration
NO

#### Description
WIP for issue #1996. Not tested yet.

### CODE REVIEW

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            _configService = configService;
+            _logger = logger;
+        }
+
+        public SpecificationPriority Priority => SpecificationPriority.Default;
+        public RejectionType Type => RejectionType.Permanent;
+
+        public Decision IsSatisfiedBy(RemoteEpisode subject, SearchCriteriaBase searchCriteria)
+        {
+            var size = subject.Release.Size;
```
Comment (margaale) [inline src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: In what unit is this size?

Comment (Taloth) [reply src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: bytes, always bytes.

File: src/NzbDrone.Core.Test/DecisionEngineTests/MaximumSizeSpecificationFixture.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        [Test]
+        public void should_return_true_when_size_is_smaller_than_maximum_size()
+        {
+            WithMaximumSize(2000);
+            WithSize(1999);
+
+            Subject.IsSatisfiedBy(_remoteEpisode, null).Accepted.Should().BeTrue();
+        }
+
+        public void should_return_true_when_size_is_equal_than_maximum_size()
```
Comment (Taloth) [inline src/NzbDrone.Core.Test/DecisionEngineTests/MaximumSizeSpecificationFixture.cs]: Missing `[Test]`

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs (hunk 1 above)
Comment (margaale) [reply src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: Thanks, is there a helper to convert to/from MB? or i should do it by dividing/multiplying by hand?

Comment (Taloth) [reply src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: `_configService.MaximumSize.Megabytes()` although that's actually Mebibytes.

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                _logger.Debug("Maximum size is not set.");
+                return Decision.Accept();
+            }
+
+            _logger.Debug("Checking if report meets maximum size requirements. {0}", size);
+
+            if (size > maximumSize)
+            {
+                var message = $"{size} too big, maximumn size is {maximumSize} MB";
```
Comment (Taloth) [inline src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: use `.SizeSuffix()` in both cases.

**Comment by Taloth** [comment]
Forgot to mention that releases can have zero size.

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        private readonly Logger _logger;
+
+        public MaximumSizeSpecification(IConfigService configService, Logger logger)
+        {
+            _configService = configService;
+            _logger = logger;
+        }
+
+        public SpecificationPriority Priority => SpecificationPriority.Default;
+        public RejectionType Type => RejectionType.Permanent;
```
Comment (margaale) [inline src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: What's the criteria for setting those two properties? I just guessed those values...

File: src/NzbDrone.Api/Config/IndexerConfigModule.cs [hunk 5]
```diff
@@ -13,6 +13,9 @@ public IndexerConfigModule(IConfigService configService)
             SharedValidator.RuleFor(c => c.MinimumAge)
                            .GreaterThanOrEqualTo(0);
 
+            SharedValidator.RuleFor(c => c.MaximumSize)
+                            .GreaterThanOrEqualTo(0);
```
Comment (Taloth) [inline src/NzbDrone.Api/Config/IndexerConfigModule.cs]: Nitpick: one space too much.
Wouldn't have commented on it if you didn't have to change something else. :smile:

**Comment by margaale** [comment]
But it is passing, or should be as I didn't try it, with this logic.

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs (hunk 4 above)
Comment (margaale) [reply src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: So my guessing was correct... Thanks for the explanation

**Comment by Taloth** [comment]
of course, stupid of me coz 0 < any max :smile:

File: src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                _logger.Debug("Maximum size is not set.");
+                return Decision.Accept();
+            }
+
+            _logger.Debug("Checking if report meets maximum size requirements. {0}", size.SizeSuffix());
+
+            if (size > maximumSize)
+            {
+                var message = $"{size.SizeSuffix()} is too big, maximumn size is {maximumSize.SizeSuffix()}";
```
Comment (Taloth) [inline src/NzbDrone.Core/DecisionEngine/Specifications/MaximumSizeSpecification.cs]: Noticed this typo just before I was about to press the merge button.

**Comment by margaale** [comment]
Do you want me to squash this?

**Comment by margaale** [comment]
> of course, stupid of me coz 0 < any max 😄

I added the code to explicitly log that escenario

**Comment by Safihre** [comment]
🎉❤️

**Pull request merged**